WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,147 @@
+                                                                   configuration: CardPresentConfigurationLoader().configuration,
+                                                                   alertsPresenter: self,
+                                                                   preflightController: preflightController)
+        return await withCheckedContinuation { continuation in
+            orderPaymentUseCase.collectPayment(using: discoveryMethod) { error in
+                if let error = error as? CardPaymentErrorProtocol {
+                    continuation.resume(returning: .failure(error))
+                } else {
+                    continuation.resume(returning: .failure(CardPaymentsAdaptorError.unknownPaymentError(underlyingError: error)))
+                }
+            } onCancel: {
+                continuation.resume(returning: .cancellation)
+            } onPaymentCompletion: {
+                // no-op – not used in PaymentMethodsViewModel anyway so this can be removed
+            } onCompleted: {
+                continuation.resume(returning: .success(order))
+            }
+        }
+    }
+
+    enum CardPaymentsAdaptorError: Error, CardPaymentErrorProtocol {
+        var retryApproach: CardPaymentRetryApproach {
+            .restart
+        }
+
+        case unknownPaymentError(underlyingError: Error)
+    }
+}
+
+extension CardPresentPaymentsAdaptor: CardPresentPaymentsOnboardingPresenting, CardPresentPaymentAlertsPresenting {
+    func showOnboardingIfRequired(from: UIViewController, readyToCollectPayment: @escaping () -> Void) {
+        paymentsAlertHandler?.showOnboarding()
+    }
+
+    func refresh() {
+        // TODO: Refresh onboarding
+    }
+
+    func present(viewModel: CardPresentPaymentsModalViewModel) {
+        paymentsAlertHandler?.present(CardPresentPaymentsAdaptorPaymentAlert(from: viewModel))
+    }
+
+    func foundSeveralReaders(readerIDs: [String], connect: @escaping (String) -> Void, cancelSearch: @escaping () -> Void) {
+        paymentsAlertHandler?.showReaderList(readerIDs)
+        // the button actions here might need to be communicated... or we could expose them on the adaptor somehow.
+    }
+
+    func updateSeveralReadersList(readerIDs: [String]) {
+        paymentsAlertHandler?.showReaderList(readerIDs)
+    }
+
+    func dismiss() {
+        paymentsAlertHandler?.dismiss()
+        // We might not really need this
+    }
+}
+
+enum CardPresentPaymentEvent {
+    case presentAlert(CardPresentPaymentsAdaptorPaymentAlert)
+    case presentReaderList(_ readerIDs: [String])
+    case showOnboarding
+}
+
+struct CardPresentPaymentsAdaptorPaymentAlert {
+    init(from paymentsModalViewModel: CardPresentPaymentsModalViewModel) {
+        // In here we still need to handle the button actions wrt the UIViewControllers the closures are passed.
+        // That said, very few of the alerts actually use the UIVCs, so it might be just as easy to remove the dependency from both sides
```

Comment:
We should also update (or remove) `var image: UIImage { get }` property since uses a specific UIKit image type. The `auxiliaryAttributedButtonTitle` property is not a problem for SwiftUI, but it's used only once through all modals, so maybe is worth take a look and see if we should remove it as well for simplicity.

Reply:
Yeah, I'm not totally sure how much we want to carry across from the `CardPresentPaymentsModalViewModel` here, but the nice thing is that we don't need to include the image if we don't want to. We can handle that however we choose within the consuming code – perhaps the image is the same for all of them, perhaps we determine it based on the title, or we add an enum somewhere for translating (back?) to a more declarative type of output.